File: solutions/julia/grains/1/grains.jl
"""Calculate the number of grains on square `square`."""
function on_square(square)
    n::UInt128=1
    if square <= 0
        throw(DomainError(square))
    end
    if square > 64 0
        throw(DomainError(square))
    end
    for i in 2:square
        n=n*2
    end
    return n
end

"""Calculate the total number of grains after square `square`."""
function total_after(square)
    s::UInt128=0
    if square <= 0
        throw(DomainError(square))
    end
    if square > 64 0
        throw(DomainError(square))
    end
    for i in 1:square
        s=s+on_square(i)
    end
    return s
end

File: solutions/javascript/grains/1/grains.js
/**
 * You can use the bigint type and BigInt global object to support numbers below
 * Number.MIN_SAFE_INTEGER and above NUMBER.MAX_SAFE_INTEGER.
 *
 * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
 */

//
// This is only a SKELETON file for the 'Grains' exercise. It's been provided as a
// convenience to get you started writing code faster.
//
export const square = (n) => {
  if (n < 1 || n > 64) throw new Error('square must be between 1 and 64');
  var sn = BigInt(1);
  for (var i=2; i<=n; i++){
    sn = sn * BigInt(2);
  }
  return sn;
};

export const total = () => {
  var s = BigInt(0);
  for (var i=1; i<= 64; i++) {
    s += square(i);
  }
  return s;
};
